File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/voicechat/VoiceMessages.java
package crabcraft.net.crabUtilities.voicechat;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.UUID;

/**
 * Wire format for the Redis voice bus.
 *
 * <p>Audio frames on {@code crabcraft:svc:audio:&lt;uuid&gt;} are
 * length-prefixed binary because the opus payload is not text.
 *
 * <p>Roster lifecycle messages on {@code crabcraft:svc:roster} are
 * NUL-separated UTF-8 strings, the first field being the opcode. We
 * carry no profile data — cross-server group members rely on the
 * existing tab-list sync to populate the receiving client's
 * {@code playerInfoMap}, which is what SVC reads from for skins.
 */
final class VoiceMessages {

    static final String AUDIO_CHANNEL_PREFIX = "crabcraft:svc:audio:";
    static final String PLAYER_HOME_KEY_PREFIX = "crabcraft:svc:player-home:";
    static final String PLAYER_GROUP_KEY_PREFIX = "crabcraft:svc:player-group:";
    static final String CALL_METADATA_KEY_PREFIX = "crabcraft:svc:call:";
    static final String CALL_TARGET_KEY_PREFIX = "crabcraft:svc:call-target:";
    static final String ROSTER_CHANNEL = "crabcraft:svc:roster";

    static final String SEP = "\0";

    static final String OP_ROSTER_JOIN = "ROSTER_JOIN";
    static final String OP_ROSTER_LEAVE = "ROSTER_LEAVE";
    static final String OP_CALL_JOIN = "CALL_JOIN";
    static final String OP_CALL_RING_START = "CALL_RING_START";
    static final String OP_CALL_RING_STOP = "CALL_RING_STOP";

    private VoiceMessages() {}

    static String audioChannel(UUID groupId) {
        return AUDIO_CHANNEL_PREFIX + groupId;
    }

    static String playerHomeKey(UUID playerId) {
        return PLAYER_HOME_KEY_PREFIX + playerId;
    }

    static String playerGroupKey(UUID playerId) {
        return PLAYER_GROUP_KEY_PREFIX + playerId;
    }

    static String callMetadataKey(UUID groupId) {
        return CALL_METADATA_KEY_PREFIX + groupId;
    }

    static String callTargetKey(UUID playerId) {
        return CALL_TARGET_KEY_PREFIX + playerId;
    }

    /* ----------------------- Audio ----------------------- */

    /**
     * Audio frame layout (binary):
     * <pre>
     *   2 bytes: backend name length (uint16)
     *   N bytes: backend name (UTF-8)
     *  16 bytes: speaker UUID
     *   1 byte:  whispering flag (0/1)
     *   2 bytes: opus length (uint16)
     *   N bytes: opus data
     * </pre>
     */
    static byte[] encodeAudioFrame(String backend, UUID speaker,
                                   boolean whispering, byte[] opus) {
        byte[] backendBytes = backend.getBytes(StandardCharsets.UTF_8);
        ByteBuffer buf = ByteBuffer.allocate(2 + backendBytes.length + 16 + 1 + 2 + opus.length);
        buf.putShort((short) backendBytes.length);
        buf.put(backendBytes);
        buf.putLong(speaker.getMostSignificantBits());
        buf.putLong(speaker.getLeastSignificantBits());
        buf.put((byte) (whispering ? 1 : 0));
        buf.putShort((short) opus.length);
        buf.put(opus);
        return buf.array();
    }

    static AudioFrame decodeAudioFrame(byte[] data) {
        ByteBuffer buf = ByteBuffer.wrap(data);
        int backendLen = buf.getShort() & 0xFFFF;
        byte[] backendBytes = new byte[backendLen];
        buf.get(backendBytes);
        String backend = new String(backendBytes, StandardCharsets.UTF_8);
        UUID speaker = new UUID(buf.getLong(), buf.getLong());
        boolean whispering = buf.get() != 0;
        int opusLen = buf.getShort() & 0xFFFF;
        byte[] opus = new byte[opusLen];
        buf.get(opus);
        return new AudioFrame(backend, speaker, whispering, opus);
    }

    record AudioFrame(String homeBackend, UUID speaker, boolean whispering, byte[] opus) {}

    /* ----------------------- Roster lifecycle ----------------------- */

    /**
     * {@code ROSTER_JOIN<NUL>group<NUL>player<NUL>name<NUL>backend}
     */
    static String encodeRosterJoin(UUID groupId, UUID playerId, String name, String backend) {
        return String.join(SEP, OP_ROSTER_JOIN, groupId.toString(),
                playerId.toString(), name == null ? "" : name, backend);
    }

    static RosterJoin decodeRosterJoin(String message) {
        String[] parts = message.split(SEP, -1);
        if (parts.length < 5) return null;
        try {
            return new RosterJoin(UUID.fromString(parts[1]),
                    UUID.fromString(parts[2]), parts[3], parts[4]);
        } catch (IllegalArgumentException e) {
            return null;
        }
    }

    record RosterJoin(UUID groupId, UUID playerId, String name, String backend) {}

    static String encodeRosterLeave(UUID groupId, UUID playerId, String backend) {
        return String.join(SEP, OP_ROSTER_LEAVE, groupId.toString(),
                playerId.toString(), backend);
    }

    static RosterLeave decodeRosterLeave(String message) {
        String[] parts = message.split(SEP, -1);
        if (parts.length < 4) return null;
        try {
            return new RosterLeave(UUID.fromString(parts[1]),
                    UUID.fromString(parts[2]), parts[3]);
        } catch (IllegalArgumentException e) {
            return null;
        }
    }

    record RosterLeave(UUID groupId, UUID playerId, String backend) {}

    /**
     * {@code CALL_JOIN<NUL>group<NUL>player<NUL>password}
     */
    static String encodeCallJoin(UUID groupId, UUID playerId, String password) {
        if (!isValidCallPassword(password)) {
            throw new IllegalArgumentException("Call password must be a URL-safe 22-64 character secret");
        }
        return String.join(SEP, OP_CALL_JOIN, groupId.toString(),
                playerId.toString(), password);
    }

    static CallJoin decodeCallJoin(String message) {
        if (message == null) return null;
        String[] parts = message.split(SEP, -1);
        if (parts.length != 4 || !OP_CALL_JOIN.equals(parts[0])
                || !isValidCallPassword(parts[3])) return null;
        try {
            return new CallJoin(UUID.fromString(parts[1]),
                    UUID.fromString(parts[2]), parts[3]);
        } catch (IllegalArgumentException e) {
            return null;
        }
    }

    static boolean isValidCallPassword(String password) {
        if (password == null || password.length() < 22 || password.length() > 64) return false;
        for (int i = 0; i < password.length(); i++) {
            char character = password.charAt(i);
            if (!(character >= 'a' && character <= 'z')
                    && !(character >= 'A' && character <= 'Z')
                    && !(character >= '0' && character <= '9')
                    && character != '-' && character != '_') return false;
        }
        return true;
    }

    record CallJoin(UUID groupId, UUID playerId, String password) {}

    /**
     * {@code CALL_RING_START<NUL>token<NUL>player<NUL>direction<NUL>expiresAtMillis}
     */
    static String encodeCallRingStart(String token, UUID playerId,
                                      RingDirection direction, long expiresAtMillis) {
        if (!isValidCallRingToken(token) || expiresAtMillis <= 0L) {
            throw new IllegalArgumentException("Invalid call ringtone start");
        }
        return String.join(SEP, OP_CALL_RING_START, token, playerId.toString(),
                direction.name(), Long.toString(expiresAtMillis));
    }

    static CallRingStart decodeCallRingStart(String message) {
        if (message == null) return null;
        String[] parts = message.split(SEP, -1);
        if (parts.length != 5 || !OP_CALL_RING_START.equals(parts[0])
                || !isValidCallRingToken(parts[1])) return null;
        try {
            UUID playerId = UUID.fromString(parts[2]);
            RingDirection direction = RingDirection.valueOf(parts[3]);
            long expiresAtMillis = Long.parseLong(parts[4]);
            if (!playerId.toString().equals(parts[2]) || expiresAtMillis <= 0L
                    || !Long.toString(expiresAtMillis).equals(parts[4])) return null;
            return new CallRingStart(parts[1], playerId, direction, expiresAtMillis);
        } catch (IllegalArgumentException e) {
            return null;
        }
    }

    /**
     * {@code CALL_RING_STOP<NUL>token<NUL>player<NUL>direction}
     */
    static String encodeCallRingStop(String token, UUID playerId, RingDirection direction) {
        if (!isValidCallRingToken(token)) {
            throw new IllegalArgumentException("Invalid call ringtone stop");
        }
        return String.join(SEP, OP_CALL_RING_STOP, token, playerId.toString(), direction.name());
    }

    static CallRingStop decodeCallRingStop(String message) {
        if (message == null) return null;
        String[] parts = message.split(SEP, -1);
        if (parts.length != 4 || !OP_CALL_RING_STOP.equals(parts[0])
                || !isValidCallRingToken(parts[1])) return null;
        try {
            UUID playerId = UUID.fromString(parts[2]);
            if (!playerId.toString().equals(parts[2])) return null;
            return new CallRingStop(parts[1], playerId, RingDirection.valueOf(parts[3]));
        } catch (IllegalArgumentException e) {
            return null;
        }
    }

    static boolean isValidCallRingToken(String token) {
        if (token == null || token.length() < 22 || token.length() > 64) return false;
        for (int i = 0; i < token.length(); i++) {
            char character = token.charAt(i);
            if (!(character >= 'a' && character <= 'z')
                    && !(character >= 'A' && character <= 'Z')
                    && !(character >= '0' && character <= '9')
                    && character != '-' && character != '_') return false;
        }
        return true;
    }

    enum RingDirection {
        INCOMING,
        OUTGOING
    }

    record CallRingStart(String token, UUID playerId, RingDirection direction,
                         long expiresAtMillis) {}

    record CallRingStop(String token, UUID playerId, RingDirection direction) {}
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/voicechat/VoiceMessages.kt
package crabcraft.net.crabUtilities.voicechat

import de.maxhenkel.voicechat.api.Group
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets.UTF_8
import java.util.Base64
import java.util.UUID

/** Binary audio and NUL-separated lifecycle messages for the Redis voice bus. */
object VoiceMessages {
    const val AUDIO_CHANNEL_PREFIX = "crabcraft:svc:audio:"
    const val PLAYER_HOME_KEY_PREFIX = "crabcraft:svc:player-home:"
    const val PLAYER_GROUP_KEY_PREFIX = "crabcraft:svc:player-group:"
    const val CALL_TARGET_KEY_PREFIX = "crabcraft:svc:call-target:"
    const val GROUP_MEMBERS_KEY_PREFIX = "crabcraft:svc:group-members:"
    const val GROUPS_REGISTRY_KEY = "crabcraft:svc:groups"
    const val PERMANENT_GROUPS_KEY = "crabcraft:svc:groups:permanent"
    const val ROSTER_CHANNEL = "crabcraft:svc:roster"
    const val LIFECYCLE_CHANNEL = "crabcraft:svc:lifecycle"
    const val SEP = "\u0000"
    const val OP_GROUP_CHANGED = "GROUP_CHANGED"
    const val OP_ROSTER_JOIN = "ROSTER_JOIN"
    const val OP_ROSTER_LEAVE = "ROSTER_LEAVE"
    const val OP_CALL_JOIN = "CALL_JOIN"
    const val OP_CALL_RING_START = "CALL_RING_START"
    const val OP_CALL_RING_STOP = "CALL_RING_STOP"

    @JvmStatic fun audioChannel(groupId: UUID) = AUDIO_CHANNEL_PREFIX + groupId

    @JvmStatic fun playerHomeKey(playerId: UUID) = PLAYER_HOME_KEY_PREFIX + playerId

    @JvmStatic fun playerGroupKey(playerId: UUID) = PLAYER_GROUP_KEY_PREFIX + playerId

    @JvmStatic fun callTargetKey(playerId: UUID) = CALL_TARGET_KEY_PREFIX + playerId

    @JvmStatic fun routeBackend(route: String?): String? = route?.substringBefore(SEP)

    /** uint16 route length, UTF-8 route, UUID, whisper flag, uint16 opus length, opus. */
    @JvmStatic
    fun encodeAudioFrame(route: String, speaker: UUID, whispering: Boolean, opus: ByteArray): ByteArray {
        val routeBytes = route.toByteArray(UTF_8)
        return ByteBuffer.allocate(2 + routeBytes.size + 16 + 1 + 2 + opus.size)
            .putShort(routeBytes.size.toShort())
            .put(routeBytes)
            .putLong(speaker.mostSignificantBits)
            .putLong(speaker.leastSignificantBits)
            .put(if (whispering) 1.toByte() else 0.toByte())
            .putShort(opus.size.toShort())
            .put(opus)
            .array()
    }

    @JvmStatic
    fun decodeAudioFrame(data: ByteArray): AudioFrame {
        val buf = ByteBuffer.wrap(data)
        val routeBytes = ByteArray(buf.short.toInt() and 0xFFFF).also(buf::get)
        val speaker = UUID(buf.long, buf.long)
        val whispering = buf.get() != 0.toByte()
        val opus = ByteArray(buf.short.toInt() and 0xFFFF).also(buf::get)
        return AudioFrame(String(routeBytes, UTF_8), speaker, whispering, opus)
    }

    data class AudioFrame(val route: String, val speaker: UUID, val whispering: Boolean, val opus: ByteArray) {
        fun route() = route

        fun speaker() = speaker

        fun whispering() = whispering

        fun opus() = opus
    }

    /** Passwords stay in the authoritative Redis hash; pub/sub only invalidates an ID. */
    @JvmStatic
    fun encodeGroupDefinition(group: GroupDefinition): String =
        listOf(
                encodeText(group.name),
                if (group.password == null) "0" else "1",
                group.password?.let(::encodeText) ?: "",
                typeToString(group.type),
                if (group.hidden) "1" else "0",
                if (group.permanent) "1" else "0",
            )
            .joinToString(SEP)

    @JvmStatic
    fun decodeGroupDefinition(id: UUID, encoded: String?): GroupDefinition? {
        val parts = encoded?.split(SEP) ?: return null
        if (parts.size != 6) return null
        return try {
            GroupDefinition(
                id,
                decodeText(parts[0]),
                if (parts[1] == "1") decodeText(parts[2]) else null,
                typeFromString(parts[3]),
                parts[4] == "1",
                parts[5] == "1",
            )
        } catch (_: IllegalArgumentException) {
            null
        }
    }

    @JvmStatic fun encodeGroupChanged(groupId: UUID) = listOf(OP_GROUP_CHANGED, groupId.toString()).joinToString(SEP)

    @JvmStatic
    fun decodeGroupChanged(message: String): UUID? {
        val parts = message.split(SEP)
        if (parts.size != 2 || parts[0] != OP_GROUP_CHANGED) return null
        return try {
            UUID.fromString(parts[1])
        } catch (_: IllegalArgumentException) {
            null
        }
    }

    @JvmStatic
    fun typeToString(type: Group.Type): String =
        when (type) {
            Group.Type.NORMAL -> "NORMAL"
            Group.Type.OPEN -> "OPEN"
            Group.Type.ISOLATED -> "ISOLATED"
            else -> throw IllegalArgumentException("Unknown group type")
        }

    @JvmStatic
    fun typeFromString(value: String): Group.Type =
        when (value) {
            "NORMAL" -> Group.Type.NORMAL
            "OPEN" -> Group.Type.OPEN
            "ISOLATED" -> Group.Type.ISOLATED
            else -> throw IllegalArgumentException("Unknown group type")
        }

    private fun encodeText(value: String) = Base64.getEncoder().encodeToString(value.toByteArray(UTF_8))

    private fun decodeText(value: String) = String(Base64.getDecoder().decode(value), UTF_8)

    data class GroupDefinition(
        val id: UUID,
        val name: String,
        val password: String?,
        val type: Group.Type,
        val hidden: Boolean,
        val permanent: Boolean,
    ) {
        fun id() = id

        fun name() = name

        fun password() = password

        fun type() = type

        fun hidden() = hidden

        fun permanent() = permanent
    }

    @JvmStatic
    fun encodeRosterJoin(groupId: UUID, playerId: UUID, name: String?, route: String) =
        listOf(OP_ROSTER_JOIN, groupId.toString(), playerId.toString(), name ?: "", route).joinToString(SEP)

    @JvmStatic
    fun decodeRosterJoin(message: String): RosterJoin? {
        val parts = message.split(SEP, limit = 5)
        if (parts.size != 5 || parts[0] != OP_ROSTER_JOIN) return null
        return try {
            RosterJoin(UUID.fromString(parts[1]), UUID.fromString(parts[2]), parts[3], parts[4])
        } catch (_: IllegalArgumentException) {
            null
        }
    }

    data class RosterJoin(val groupId: UUID, val playerId: UUID, val name: String, val route: String) {
        fun groupId() = groupId

        fun playerId() = playerId

        fun name() = name

        fun route() = route

        fun backend() = routeBackend(route)!!
    }

    @JvmStatic
    fun encodeRosterLeave(groupId: UUID, playerId: UUID, route: String) =
        listOf(OP_ROSTER_LEAVE, groupId.toString(), playerId.toString(), route).joinToString(SEP)

    @JvmStatic
    fun decodeRosterLeave(message: String): RosterLeave? {
        val parts = message.split(SEP, limit = 4)
        if (parts.size != 4 || parts[0] != OP_ROSTER_LEAVE) return null
        return try {
            RosterLeave(UUID.fromString(parts[1]), UUID.fromString(parts[2]), parts[3])
        } catch (_: IllegalArgumentException) {
            null
        }
    }

    data class RosterLeave(val groupId: UUID, val playerId: UUID, val route: String) {
        fun groupId() = groupId

        fun playerId() = playerId

        fun route() = route

        fun backend() = routeBackend(route)!!
    }

    /** Password-free wake-up hint; the target and group secret are read from Redis. */
    @JvmStatic
    fun encodeCallJoin(groupId: UUID, playerId: UUID, generation: String): String {
        requireOpaqueToken(generation)
        return listOf(OP_CALL_JOIN, groupId.toString(), playerId.toString(), generation).joinToString(SEP)
    }

    @JvmStatic
    fun decodeCallJoin(message: String): CallJoin? {
        val parts = message.split(SEP)
        if (parts.size != 4 || parts[0] != OP_CALL_JOIN || !isOpaqueToken(parts[3])) return null
        return try {
            CallJoin(parseCanonicalUuid(parts[1]), parseCanonicalUuid(parts[2]), parts[3])
        } catch (_: IllegalArgumentException) {
            null
        }
    }

    data class CallJoin(val groupId: UUID, val playerId: UUID, val generation: String) {
        fun groupId() = groupId

        fun playerId() = playerId

        fun generation() = generation

        fun target() = CallTarget(groupId, generation)
    }

    @JvmStatic
    fun encodeCallTarget(target: CallTarget): String {
        requireOpaqueToken(target.generation)
        return listOf(target.groupId.toString(), target.generation).joinToString(SEP)
    }

    @JvmStatic
    fun decodeCallTarget(encoded: String?): CallTarget? {
        val parts = encoded?.split(SEP) ?: return null
        if (parts.size != 2 || !isOpaqueToken(parts[1])) return null
        return try {
            CallTarget(parseCanonicalUuid(parts[0]), parts[1])
        } catch (_: IllegalArgumentException) {
            null
        }
    }

    data class CallTarget(val groupId: UUID, val generation: String) {
        fun groupId() = groupId

        fun generation() = generation
    }

    @JvmStatic
    fun encodeCallRingStart(token: String?, playerId: UUID?, direction: RingDirection?, expiresAtMillis: Long): String {
        requireOpaqueToken(token)
        require(playerId != null && direction != null && expiresAtMillis > 0) { "Invalid call ringtone start" }
        return listOf(OP_CALL_RING_START, token, playerId.toString(), direction.name, expiresAtMillis.toString())
            .joinToString(SEP)
    }

    @JvmStatic
    fun decodeCallRingStart(message: String): CallRingStart? {
        val parts = message.split(SEP)
        if (parts.size != 5 || parts[0] != OP_CALL_RING_START || !isOpaqueToken(parts[1])) return null
        return try {
            CallRingStart(
                parts[1],
                parseCanonicalUuid(parts[2]),
                RingDirection.valueOf(parts[3]),
                parseCanonicalPositiveLong(parts[4]),
            )
        } catch (_: IllegalArgumentException) {
            null
        }
    }

    data class CallRingStart(
        val token: String,
        val playerId: UUID,
        val direction: RingDirection,
        val expiresAtMillis: Long,
    ) {
        fun token() = token

        fun playerId() = playerId

        fun direction() = direction

        fun expiresAtMillis() = expiresAtMillis
    }

    @JvmStatic
    fun encodeCallRingStop(token: String?, playerId: UUID?, direction: RingDirection?): String {
        requireOpaqueToken(token)
        require(playerId != null && direction != null) { "Invalid call ringtone stop" }
        return listOf(OP_CALL_RING_STOP, token, playerId.toString(), direction.name).joinToString(SEP)
    }

    @JvmStatic
    fun decodeCallRingStop(message: String): CallRingStop? {
        val parts = message.split(SEP)
        if (parts.size != 4 || parts[0] != OP_CALL_RING_STOP || !isOpaqueToken(parts[1])) return null
        return try {
            CallRingStop(parts[1], parseCanonicalUuid(parts[2]), RingDirection.valueOf(parts[3]))
        } catch (_: IllegalArgumentException) {
            null
        }
    }

    data class CallRingStop(val token: String, val playerId: UUID, val direction: RingDirection) {
        fun token() = token

        fun playerId() = playerId

        fun direction() = direction
    }

    enum class RingDirection {
        INCOMING,
        OUTGOING,
    }

    private fun requireOpaqueToken(token: String?) {
        require(isOpaqueToken(token)) { "Invalid call token" }
    }

    private fun isOpaqueToken(token: String?): Boolean =
        token != null &&
            token.length in 22..64 &&
            token.all { it in 'a'..'z' || it in 'A'..'Z' || it in '0'..'9' || it == '-' || it == '_' }

    private fun parseCanonicalUuid(value: String): UUID =
        UUID.fromString(value).also {
            require(it.toString() == value) { "Non-canonical UUID" }
        }

    private fun parseCanonicalPositiveLong(value: String): Long =
        value.toLong().also {
            require(it > 0 && it.toString() == value) { "Non-canonical positive long" }
        }
}
